internal/server/v1/optimus/service.go

```diff
@@ -36,13 +47,56 @@ func (svc *Service) FindJobSpec(ctx context.Context, jobName, projectName string
 	return list[0], nil
 }
 
-func (svc *Service) ListJobs(ctx context.Context, projectName string) ([]*optimusv1beta1.JobSpecificationResponse, error) {
-	res, err := svc.client.GetJobSpecifications(ctx, &optimusv1beta1.GetJobSpecificationsRequest{
-		ProjectName: projectName,
+func (svc *Service) ListJobs(ctx context.Context, projectSlug string) ([]*optimusv1beta1.JobSpecificationResponse, error) {
+	optimusCl, err := svc.getOptimusClient(ctx, projectSlug)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := optimusCl.GetJobSpecifications(ctx, &optimusv1beta1.GetJobSpecificationsRequest{
+		ProjectName: projectSlug,
 	})
 	if err != nil {
 		return nil, err
 	}
 
 	return res.JobSpecificationResponses, nil
 }
+
+func (svc *Service) getOptimusClient(ctx context.Context, projectSlug string) (optimusv1beta1grpc.JobSpecificationServiceClient, error) {
+	// retrieve hostname from cache
+
+	if cl, exists := svc.cache.data[projectSlug]; exists {
+		return cl, nil
+	}
+	// retrieve hostname from shield
+	prj, err := svc.shieldClient.GetProject(ctx, &shieldv1beta1.GetProjectRequest{
+		Id: projectSlug,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	metadata := prj.Project.Metadata.AsMap()
+	optimusHost, exists := metadata[optimusHostKey]
+	if !exists {
+		return nil, ErrOptimusHostNotFound
+	}
+
+	optimusHostStr, isString := optimusHost.(string)
+	if !isString {
+		return nil, ErrOptimusHostNotString
+	}
+
+	cl, err := svc.builder.BuildOptimusClient(optimusHostStr)
+	if err != nil {
+		return nil, err
+	}
+
+	// store hostname in cache
+	svc.cache.mu.Lock()
```

Comment:
if you are handling the mutex here, we should might as well remove `Cache` struct and add the mutex to `Service` struct instead.

If we are going with `Cache` struct, mutex locking should be handled there.

it could have methods like
```go
Get(key string) (client JobSpecClient, exist bool)
Set(key string, client JobSpecClient)
```

Reply:
done